File: contents/code/effects.py
#!/usr/bin/python2 -B
# -*- coding: utf-8 -*-
#
# Webcamod, webcam capture plasmoid.
# Copyright (C) 2011-2012  Gonzalo Exequiel Pedone
#
# Webcamod is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Webcamod is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Webcamod. If not, see <http://www.gnu.org/licenses/>.
#
# Email     : hipersayan DOT x AT gmail DOT com
# Web-Site 1: http://github.com/hipersayanX/Webcamoid
# Web-Site 2: http://kde-apps.org/content/show.php/Webcamoid?content=144796

import os
import sys

from PyQt4 import QtCore, QtGui, uic
from PyKDE4 import kdeui
import v4l2tools

import appenvironment


class Effects(QtGui.QWidget):
    def __init__(self, parent=None, tools=None):
        QtGui.QWidget.__init__(self, parent)

        self.appEnvironment = appenvironment.AppEnvironment(self)
        uic.loadUi(self.resolvePath('../ui/effects.ui'), self)

        self.setWindowIcon(kdeui.KIcon('camera-web'))
        self.btnUp.setIcon(kdeui.KIcon('arrow-up'))
        self.btnAdd.setIcon(kdeui.KIcon('arrow-right'))
        self.btnDown.setIcon(kdeui.KIcon('arrow-down'))
        self.btnRemove.setIcon(kdeui.KIcon('arrow-left'))
        self.btnReset.setIcon(kdeui.KIcon('edit-undo'))

        self.tools = tools if tools else v4l2tools.V4L2Tools(self, True)
        self.effects = []
        effects = self.tools.availableEffects()

        for index, effect in enumerate(effects.keys()):
            item = [effect, QtGui.QListWidgetItem(effects[effect])]
            self.effects.append(item)
            self.lswEffects.addItem(item[1])

        self.lswEffects.sortItems()

        for effect in self.tools.currentEffects():
            item, index = self.findEffect(effect)

            if item == []:
                continue

            listWidgetItem = self.lswEffects.takeItem(self.lswEffects.row(item[1]))
            self.lswApply.addItem(listWidgetItem)

        self.tools.recordingStateChanged.connect(self.recordingStateChanged)
        self.recordingStateChanged(self.tools.recording)

    def resolvePath(self, relpath=''):
        return os.path.normpath(os.path.join(os.path.
                                dirname(os.path.realpath(__file__)), relpath))

    def findEffectWidget(self, widget=None):
        for index, effect in enumerate(self.effects):
            if effect[1] == widget:
                return effect, index

        return [], -1

    def findEffect(self, effect=''):
        for index, effect_ in enumerate(self.effects):
            if effect_[0] == effect:
                return effect_, index

        return [], -1

    @QtCore.pyqtSlot(QtGui.QImage, str)
    def setEffectPreview(self, image=None, effect=''):
        item, index = self.findEffect(effect)
        item[1].setIcon(QtGui.QIcon(QtGui.QPixmap.fromImage(image)))

    @QtCore.pyqtSlot(bool)
    def recordingStateChanged(self, recording):
        self.lswEffects.setEnabled(not recording)
        self.lswApply.setEnabled(not recording)

    @QtCore.pyqtSlot(str)
    def on_txtSearch_textChanged(self, text):
        regexp = QtCore.QRegExp(text, QtCore.Qt.CaseInsensitive, QtCore.QRegExp.Wildcard)

        for i in range(self.lswEffects.count()):
            if regexp.indexIn(self.lswEffects.item(i).text()) == -1:
                self.lswEffects.item(i).setHidden(True)
            else:
                self.lswEffects.item(i).setHidden(False)

        for i in range(self.lswApply.count()):
            if regexp.indexIn(self.lswApply.item(i).text()) == -1:
                self.lswApply.item(i).setHidden(True)
            else:
                self.lswApply.item(i).setHidden(False)

    @QtCore.pyqtSlot()
    def on_btnAdd_clicked(self):
        for item in self.lswEffects.selectedItems():
            self.lswApply.addItem(self.lswEffects.
                                            takeItem(self.lswEffects.row(item)))

        self.update()

    @QtCore.pyqtSlot()
    def on_btnRemove_clicked(self):
        for item in self.lswApply.selectedItems():
            self.lswEffects.addItem(self.lswApply.
                                            takeItem(self.lswApply.row(item)))

        self.lswEffects.sortItems()
        self.update()

    @QtCore.pyqtSlot()
    def on_btnUp_clicked(self):
        for item in self.lswApply.selectedItems():
            row = self.lswApply.row(item)
            row_ = row - 1 if row >= 1 else 0
            item_ = self.lswApply.takeItem(row)
            self.lswApply.insertItem(row_, item_)
            item_.setSelected(True)

        self.update()

    @QtCore.pyqtSlot()
    def on_btnDown_clicked(self):
        for item in self.lswApply.selectedItems():
            row = self.lswApply.row(item)
            row_ = row + 1 if row < self.lswApply.count() - 1 \
                                                else self.lswApply.count() - 1
            item_ = self.lswApply.takeItem(row)
            self.lswApply.insertItem(row_, item_)
            item_.setSelected(True)

        self.update()

    @QtCore.pyqtSlot()
    def on_btnReset_clicked(self):
        while self.lswApply.count() > 0:
            self.lswEffects.addItem(self.lswApply.takeItem(0))

        self.lswEffects.sortItems()
        self.update()

    def update(self):
        if not self.tools:
            return

        effects = []

        for i in range(self.lswApply.count()):
            item, index = self.findEffectWidget(self.lswApply.item(i))
            effects.append(item[0])

        self.tools.setEffects(effects)


if __name__ == '__main__':
    app = QtGui.QApplication(sys.argv)
    effects = Effects()
    effects.show()
    app.exec_()

File: src/effects.cpp
/* Webcamoid, webcam capture application.
 * Copyright (C) 2011-2014  Gonzalo Exequiel Pedone
 *
 * Webcamod is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Webcamod is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Webcamod. If not, see <http://www.gnu.org/licenses/>.
 *
 * Email     : hipersayan DOT x AT gmail DOT com
 * Web-Site 1: http://github.com/hipersayanX/Webcamoid
 * Web-Site 2: http://opendesktop.org/content/show.php/Webcamoid?content=144796
 */

#include "ui_effects.h"

#include "effects.h"

Effects::Effects(MediaTools *mediaTools, QWidget *parent):
    QWidget(parent),
    ui(new Ui::Effects)
{
    this->ui->setupUi(this);

    this->m_mediaTools = mediaTools? mediaTools: new MediaTools(this);

    QObject::connect(this->m_mediaTools,
                     SIGNAL(deviceChanged(QString)),
                     this,
                     SLOT(deviceChanged(QString)));

    QObject::connect(this->m_mediaTools,
                     SIGNAL(effectPreviewReady(const QbPacket &)),
                     this,
                     SLOT(setEffectPreview(const QbPacket &)));

    QObject::connect(this->m_mediaTools,
                     SIGNAL(applyPreviewReady(const QbPacket &)),
                     this,
                     SLOT(setApplyPreview(const QbPacket &)));

    QObject::connect(this,
                     SIGNAL(effectVisibilityChanged(bool)),
                     this->m_mediaTools,
                     SLOT(connectPreview(bool)));

    this->m_effects = this->m_mediaTools->availableEffects();

    foreach (QString effect, this->m_effects.keys()) {
        QListWidgetItem *listWidgetItem = new QListWidgetItem(this->m_effects[effect]);

        this->m_effectsNames << effect;
        this->m_effectsWidgets << listWidgetItem;
        this->ui->lswEffects->addItem(listWidgetItem);
    }

    this->ui->lswEffects->sortItems();

    foreach (QString effect, this->m_mediaTools->currentEffects()) {
        int index = this->m_effectsNames.indexOf(effect);

        if (index < 0)
            continue;

        QListWidgetItem *listWidgetItem = this->ui->lswEffects->takeItem(this->ui->lswEffects->row(this->m_effectsWidgets[index]));
        this->ui->lswApply->addItem(listWidgetItem);
    }

    if (this->m_effects.count() > 0)
        this->ui->lswEffects->setCurrentRow(0);
}

void Effects::update()
{
   if (!this->m_mediaTools)
       return;

   QStringList effects;

   for (int i = 0; i < this->ui->lswApply->count(); i++) {
       int index = this->m_effectsWidgets.indexOf(this->ui->lswApply->item(i));

       effects << this->m_effectsNames[index];
   }

   this->m_mediaTools->setEffects(effects);
   this->ui->lswEffects->setEnabled(this->ui->lswApply->count() < 5);
}

void Effects::showEvent(QShowEvent *event)
{
    QWidget::showEvent(event);

    this->updateEffectPreview();
    emit this->effectVisibilityChanged(true);
}

void Effects::hideEvent(QHideEvent *event)
{
    QWidget::hideEvent(event);

    this->m_mediaTools->resetEffectsPreview();
    emit this->effectVisibilityChanged(false);
}

void Effects::setEffectPreview(const QbPacket &packet)
{
    if (!this->m_mediaTools->device().isEmpty()) {
        QImage image = QbUtils::packetToImage(packet);
        this->ui->lblEffectsPreview->setPixmap(QPixmap::fromImage(image));
    }
}

void Effects::setApplyPreview(const QbPacket &packet)
{
    if (!this->m_mediaTools->device().isEmpty()) {
        QImage image = QbUtils::packetToImage(packet);
        this->ui->lblApplyPreview->setPixmap(QPixmap::fromImage(image));
    }
}

void Effects::updateEffectPreview()
{
    if (this->ui->lswEffects->count() < 1)
        this->m_mediaTools->setEffectsPreview(this->m_effects.keys()[0]);
    else if (this->ui->lswEffects->currentRow() >= 0) {
        QString effectName = this->ui->lswEffects->currentItem()->text();
        this->m_mediaTools->setEffectsPreview(this->m_effects.key(effectName));
    }
}

void Effects::deviceChanged(const QString &device)
{
    if (!device.isEmpty() && this->isVisible())
        this->updateEffectPreview();
}

void Effects::on_txtSearch_textChanged(QString text)
{
    QRegExp regexp(text, Qt::CaseInsensitive, QRegExp::Wildcard);

    for (int i = 0; i < this->ui->lswEffects->count(); i++)
        if (regexp.indexIn(this->ui->lswEffects->item(i)->text()) == -1)
            this->ui->lswEffects->item(i)->setHidden(true);
        else
            this->ui->lswEffects->item(i)->setHidden(false);

    for (int i = 0; i < this->ui->lswApply->count(); i++)
        if (regexp.indexIn(this->ui->lswApply->item(i)->text()) == -1)
            this->ui->lswApply->item(i)->setHidden(true);
        else
            this->ui->lswApply->item(i)->setHidden(false);
}

void Effects::on_btnAdd_clicked()
{
    foreach (QListWidgetItem *item, this->ui->lswEffects->selectedItems())
        this->ui->lswApply->addItem(this->ui->lswEffects->
                                    takeItem(this->ui->lswEffects->row(item)));

    this->update();
}

void Effects::on_btnRemove_clicked()
{
    foreach (QListWidgetItem *item, this->ui->lswApply->selectedItems())
        this->ui->lswEffects->addItem(this->ui->lswApply->
                                      takeItem(this->ui->lswApply->row(item)));

    this->ui->lswEffects->sortItems();
    this->update();
}

void Effects::on_btnUp_clicked()
{
    foreach (QListWidgetItem *item, this->ui->lswApply->selectedItems()) {
        int row = this->ui->lswApply->row(item);
        int row_ = (row >= 1)? row - 1: 0;
        QListWidgetItem *item_ = this->ui->lswApply->takeItem(row);
        this->ui->lswApply->insertItem(row_, item_);
        item_->setSelected(true);
    }

    this->update();
}

void Effects::on_btnDown_clicked()
{
    foreach (QListWidgetItem *item, this->ui->lswApply->selectedItems()) {
        int row = this->ui->lswApply->row(item);
        int row_ = (row < this->ui->lswApply->count() - 1)? row + 1: this->ui->lswApply->count() - 1;
        QListWidgetItem *item_ = this->ui->lswApply->takeItem(row);
        this->ui->lswApply->insertItem(row_, item_);
        item_->setSelected(true);
    }

    this->update();
}

void Effects::on_btnReset_clicked()
{
    while (this->ui->lswApply->count() > 0)
        this->ui->lswEffects->addItem(this->ui->lswApply->takeItem(0));

    this->ui->lswEffects->sortItems();
    this->update();
}

void Effects::on_lswEffects_itemClicked(QListWidgetItem *item)
{
    if (!item)
        return;

    this->updateEffectPreview();
}
